Runtime/Services/BacktraceSession.cs

```diff
@@ -0,0 +1,354 @@
+﻿using Backtrace.Unity.Common;
+using Backtrace.Unity.Interfaces;
+using Backtrace.Unity.Json;
+using Backtrace.Unity.Model;
+using Backtrace.Unity.Model.JsonData;
+using Backtrace.Unity.Model.Session;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+[assembly: System.Runtime.CompilerServices.InternalsVisibleTo("Backtrace.Unity.Tests.Runtime")]
+namespace Backtrace.Unity.Services
+{
+    internal sealed class BacktraceSession : IBacktraceSession
+    {
+        /// <summary>
+        /// Startup event name that will be send on the application startup
+        /// </summary>
+        private const string StartupEventName = "Application Launches";
+
+        /// <summary>
+        /// Default number of retries
+        /// </summary>
+        public const int DefaultNumberOfRetries = 3;
+
+        /// <summary>
+        /// Time between 
+        /// </summary>
+        public const int DefaultTimeInSecBetweenRequests = 10;
+
+        /// <summary>
+        /// Maximum time between requests
+        /// </summary>
+        public const int RetryTimeMax = 5 * 60;
+
+        /// <summary>
+        /// Submission url
+        /// </summary>
+        public string SubmissionUrl { get; set; }
+
+        /// <summary>
+        /// Determine if http client should ignore ssl validation
+        /// </summary>
+        public bool IgnoreSslValidation
+        {
+            get
+            {
+                return RequestHandler.IgnoreSslValidation;
+            }
+            set
+            {
+                RequestHandler.IgnoreSslValidation = value;
+            }
+        }
+
+        private int _numberOfDroppedRequests = 0;
+
+        /// <summary>
+        /// List of unique events that will be added to next session submission payload
+        /// </summary>
+        internal readonly LinkedList<UniqueEvent> UniqueEvents = new LinkedList<UniqueEvent>();
+
+        /// <summary>
+        /// List of session events that will be added to next session submission payload
+        /// </summary>
+        internal readonly LinkedList<SessionEvent> SessionEvents = new LinkedList<SessionEvent>();
+
+        /// <summary>
+        /// Time interval in ms that algorithm uses to automatically send data to Backtrace
+        /// </summary>
+        private readonly long _timeIntervalInMs;
+
+        /// <summary>
+        /// Last update time that will be updated after each update start
+        /// </summary>
+        private float _lastUpdateTime = 0;
+
+        /// <summary>
+        /// Maximum number of events in store. If number of events in store hit the limit
+        /// BacktraceSession instance will send data to Backtrace.
+        /// </summary>
+        private uint _maximumNumberOfEventsInStore;

```

Comment:
Should Unique and Session events share the same max?

Reply:
_maximumNumberOfEventsInStore = number of events available in unique events store and session events store. For reference let me copy paste here a ShouldProcessEvent function:
```
 private bool ShouldProcessEvent(string name)
        {
            return !string.IsNullOrEmpty(name) && (_maximumNumberOfEventsInStore == 0 || (Count() + 1 <= _maximumNumberOfEventsInStore));
        }
```
and Count implementation
```
     public int Count()
        {
            return UniqueEvents.Count + SessionEvents.Count;
        }

```
as you can see I'm comparing number of events in both queues with this variable.

Reply:
I understand. I just wasn't sure if maybe Unique events and Session events should have their own maxes, like
`_maximumNumberOfUniqueEvents` and `_maximumNumberOfSessionEvents`? Maybe user wants to set different limits for different event types? Maybe we don't care about that use case though which is fine.

Reply:
I think we don't care about this use case since we send them together and this option is mostly to force sending data when we hit the limit.